### PR TITLE
Unshim ExclusiveModeGpuDiscoveryPlugin

### LINKED ISSUE #3589: [BUG] Latest 21.10 snapshot jars failing with java.lang.ClassNotFoundException: com.nvidia.spark.ExclusiveModeGpuDiscoveryPlugin
**Describe the bug**
The latest snapshot jars from 09/18 for v21.10 are failing with the following error:

```
21/09/21 21:28:25 ERROR YarnCoarseGrainedExecutorBackend [dispatcher-Executor]: Executor self-exiting due to : Unable to create executor due to com.nvidia.spark.ExclusiveModeGpuDiscoveryPlugin
java.lang.ClassNotFoundException: com.nvidia.spark.ExclusiveModeGpuDiscoveryPlugin
    at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
    at java.lang.Class.forName0(Native Method)
    at java.lang.Class.forName(Class.java:348)
    at org.apache.spark.util.Utils$.classForName(Utils.scala:207)
    at org.apache.spark.util.Utils$.$anonfun$loadExtensions$1(Utils.scala:2783)
    at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:245)
    at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
    at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
    at scala.collection.TraversableLike.flatMap(TraversableLike.scala:245)
    at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:242)
    at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
    at org.apache.spark.util.Utils$.loadExtensions(Utils.scala:2781)
    at org.apache.spark.resource.ResourceUtils$.discoverResource(ResourceUtils.scala:387)
    at org.apache.spark.resource.ResourceUtils$.$anonfun$getOrDiscoverAllResourcesForResourceProfile$3(ResourceUtils.scala:360)
    at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
    at scala.collection.MapLike$FilteredKeys.$anonfun$foreach$1(MapLike.scala:238)
    at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:234)
    at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:468)
    at scala.collection.MapLike$FilteredKeys.foreach(MapLike.scala:238)
    at scala.collection.TraversableLike.map(TraversableLike.scala:238)
    at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
    at scala.collection.AbstractTraversable.map(Traversable.scala:108)
    at org.apache.spark.resource.ResourceUtils$.getOrDiscoverAllResourcesForResourceProfile(ResourceUtils.scala:355)
    at org.apache.spark.executor.CoarseGrainedExecutorBackend.$anonfun$parseOrFindResources$2(CoarseGrainedExecutorBackend.scala:137)
    at org.apache.spark.util.Utils$.withContextClassLoader(Utils.scala:222)
    at org.apache.spark.executor.CoarseGrainedExecutorBackend.parseOrFindResources(CoarseGrainedExecutorBackend.scala:132)
    at org.apache.spark.executor.CoarseGrainedExecutorBackend.onStart(CoarseGrainedExecutorBackend.scala:93)
    at org.apache.spark.rpc.netty.Inbox.$anonfun$process$1(Inbox.scala:120)
    at org.apache.spark.rpc.netty.Inbox.safelyCall(Inbox.scala:213)
    at org.apache.spark.rpc.netty.Inbox.process(Inbox.scala:100)
    at org.apache.spark.rpc.netty.MessageLoop.org$apache$spark$rpc$netty$MessageLoop$$receiveLoop(MessageLoop.scala:75)
    at org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:41)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)

```

Related Spark config set:

```
spark.resources.discoveryPlugin com.nvidia.spark.ExclusiveModeGpuDiscoveryPlugin
```

**Steps/Code to reproduce bug**
Submit a simple Spark RAPIDS app using the latest snapshot jars from 09/18 for v21.10 

**Expected behavior**
Job should run successfully using the latest snapshot v21.10 jars.

**Environment details (please complete the following information)**
 - Environment location: YARN, Spark 3.1.2
 - Spark configuration settings related to the issue: `spark.resources.discoveryPlugin com.nvidia.spark.ExclusiveModeGpuDiscoveryPlugin`

### PR DESCRIPTION
**Pull request by gerashegalov**
Fixes #3589 

As one of the publicly documented classes loaded by Spark outside the
plugin control, ExclusiveModeGpuDiscoveryPlugin may not reside in the
Parallel World locations. Binary dedupe shows
ExclusiveModeGpuDiscoveryPlugin is bitwise-identical across all shims

Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

**Comment by gerashegalov** [comment]
```
 jar tvf dist/target/rapids-4-spark_2.12-21.10.0-SNAPSHOT.jar | grep ExclusiveModeGpu
 10734 Tue Sep 21 16:18:04 PDT 2021 com/nvidia/spark/ExclusiveModeGpuDiscoveryPlugin.class
 ```

**Comment by gerashegalov** [comment]
build

**Review by tgravescs — APPROVED** [review]

**Comment by gerashegalov** [comment]
build

**Comment by tgravescs** [comment]
we need to finish testing this manually

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala [hunk 1]
```diff
@@ -19,9 +19,9 @@ package com.nvidia.spark.rapids
 import java.net.URL
 
 import scala.collection.JavaConverters._
-
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala:22]: this needs to be put back.

**Comment by tgravescs** [comment]
weird ci hasn't reported, I don't see anything running, kicking again

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
I tried this out manually and its working for me as well:
21/09/22 06:55:01 INFO InternalExclusiveModeGpuDiscoveryPlugin: Running ExclusiveModeGpuDiscoveryPlugin to acquire 1 GPU(s), host has 1 GPU(s)

**Comment by tgravescs** [comment]
failed with the scalastyle issue because of import order I mention above

**Comment by gerashegalov** [comment]
build

**Review by tgravescs — APPROVED** [review]

**Pull request merged**